Publish v1.0.0

ssailor96 opened the pull request.

ssailor96 committed:
Working version of v1.0.0

ssailor96 committed:
Update app.py

Made changes to code to improve efficiency and readability.

ssailor96 committed:
Update README.md

README file is now up-to-date.

ssailor commented on requirements.txt:
```diff
@@ -0,0 +1 @@
+humanize==4.3.0
```
Unless you need the exact version of 4.3 you could use version specifiers to allow minor and patch versions to be used but not update to the next major version. To do that you could change this to be `humanize~=4.3`
Check out this link for more on version specifiers: https://peps.python.org/pep-0440/#examples

ssailor96 replied:
Fixed issue

ssailor commented on README.md:
```diff
@@ -1,2 +1,31 @@
+
+## Notes
+
+- Make sure to use correct input (a path to a file or directory). If incorrect input is used, the program will exit and need to be run again.
```
You may want to specify if the app requires an absolute path or a relative path.

ssailor96 replied:
I have made this change.

ssailor commented on README.md:
```diff
@@ -1,2 +1,31 @@
+
+When the program is run, it outputs the system, release, and version of the platform being used to the console. It then prompts the user to provide either the path to a file or a path to a directory. If a path to a file is provided, the program prompts the user to provide the path to a second file.
+
+To check if two files are the same, enter a path to a file. To check for duplicates in an entire directory, enter a path to a directory.
```
You may want to specify if the duplicates for the directory check checks subfolders or if it sticks to the top level.

ssailor96 replied:
Updated README to specify this.

ssailor commented on README.md:
```diff
@@ -1,2 +1,31 @@
+
+The application checks for duplicates by comparing the hashes of each file. The user is notified through the console whether or not a duplicate has been found. If any duplicates are found, the metadata, hashes, and list of duplicates for each duplicate file are saved to a file titled "duplicates_*currentdatetime*.json".
+
+## Outputs
```
It might be worth showing examples of the output for each file so users can see what the data structure of the output should look like.

ssailor96 replied:
File output and console output added to README.

ssailor commented on README.md:
```diff
@@ -1,2 +1,31 @@
+
+This application checks for duplicate files. It can compare two provided files or compare all files in a provided directory. The program creates a json file listing all duplicates as well as a json file containing metadata for the user to review.
+
+## Prerequisites
```
Since you are using an external package I'd recommend adding a small getting started section showing to create and start a virtual environment along with showing how to use the requirements.txt to download dependencies.

ssailor96 replied:
Getting started section added to README.

ssailor commented on app.py:
```diff
@@ -0,0 +1,183 @@
+# function for getting metadata for files and instantiating the objects
+
+
+def getInfo(pathList, fileList):
```
Naming conventions for python methods are all lower cases with words separated by underscores. I'd recommend changing the method names to match the naming convention. For more info here is a quick article that should help: https://www.techversantinfotech.com/python-naming-conventions-points-you-should-know/#:~:text=Method%20name%20in%20Python,it%20needs%20to%20be%20mangled.

ssailor96 replied:
Method and variable names have been changed to match python naming conventions. The only exceptions are variables being used in the output JSON file, those are named to match JSON naming conventions.

ssailor commented on app.py:
```diff
@@ -0,0 +1,183 @@
+    fileList = []
+    pathList = []
+    # find out which system, release, and version is being used
+    print("System: " + str(platform.system()))
```
You may want to consider writing this info to the JSON output files. You could create a parent object containing the run information including this nest the hashing info in a child object called results or something. Not a requirement just a thought.

ssailor96 replied:
Will consider this for future versions.

ssailor commented on app.py:
```diff
@@ -0,0 +1,183 @@
+        # get modification time and creation time for the current file
+        modTimeStamp = os.path.getmtime(filePath)
+        modTime = datetime.fromtimestamp(
+            modTimeStamp).strftime('%Y-%m-%d--%H-%M-%S')
```
I would recommend changing both the modTime and createTime to use the ISO 8601 standard. Here is some info on that https://pythontic.com/datetime/datetime/isoformat

ssailor96 replied:
Changes have been made to meet the ISO 8601 standard.

ssailor commented:
The functionality looks to be working as I would expect for this. Here are a few things I noticed that I believe would be worth considering in the future.

1. The application only works at the moment by prompting a user for input after they launch the program. This works however it does not allow for someone to be able to use this application in any form of automation. I believe this could be valuable if I wanted to perform a bunch of different hashing operations that I would trigger by using a bash script. You could change that by adding the ability for users to provide command line arguments. So as an example if someone wanted to get the hashes of a directory they could provide a command line flag with the directory name (possibly a -d flag. Then the app would run, generate the file then end which would allow the next item in a script to run. The app could be set up to still run interactively if no command line arguments are provided however if command line args are used then it would just run without having to run interactivity.

2. Since this app is meant to run on the command line I'd recommend changing the main file name to something more specific such as duplicate_checker.py 

3. I notice that 2 files are generated when you finish a hashing run but it looks like they return pretty much the same information. The duplicates file just looks like it's a shorter list of only duplicates. Since your object has a flag of isDuplicate this information could easily be parsed out. I'd recommend returning only a single file and/or giving the user the option to get the full report or just the duplicates. Either way, I think returning a single file may be more clear. 

4. After the program runs it only tells the user if there were duplicates or not and doesn't present the users with any further helpful information. I would recommend that if you're going to prompt users for input to add some more functionality after a run has been completed. This could be an option to see in the console the number of duplicates found along with a list of filenames. Additionally, other features such as asking the user if they would like to run another comparison could enhance the application as well.

ssailor commented:
There is an issue with the way you are creating file names. Colons are not allowed in filenames on windows. Using a datetime for filenames is fine but I'd recommend making sure to use characters that are safe across all OS.

ssailor reviewed (changes requested):
Need to at least fix the filename issue to use safe characters. I also recommend making some of the other changes I called out.

ssailor96 committed:
Update humanize version specifier

ssailor96 committed:
Merge pull request #2 from ssailor96/fix-requirement-versioning

Update humanize version specifier

ssailor96 committed:
Filename issues resolved.

app.py renamed to duplicate_checker.py to be more descriptive. Spaces and colons removed from output filenames to prevent errors.

ssailor96 committed:
Merge pull request #3 from ssailor96/fix-filename-issues

Filename issues resolved.

ssailor96 committed:
Output changes made.

Number of output files reduced from 2 to 1. Duplicate file names and number of duplicates are now printed to the console.

ssailor96 committed:
Output file name updated

Made suggested changes to .gitignore file. Output file name format has been updated.

ssailor96 committed:
Merge pull request #4 from ssailor96/make-output-changes

Output changes made.

ssailor96 committed:
Updated to fit naming and datetime standards

ssailor96 commented:
@ssailor 
> The functionality looks to be working as I would expect for this. Here are a few things I noticed that I believe would be worth considering in the future.
> 
> 1. The application only works at the moment by prompting a user for input after they launch the program. This works however it does not allow for someone to be able to use this application in any form of automation. I believe this could be valuable if I wanted to perform a bunch of different hashing operations that I would trigger by using a bash script. You could change that by adding the ability for users to provide command line arguments. So as an example if someone wanted to get the hashes of a directory they could provide a command line flag with the directory name (possibly a -d flag. Then the app would run, generate the file then end which would allow the next item in a script to run. The app could be set up to still run interactively if no command line arguments are provided however if command line args are used then it would just run without having to run interactivity.

I appreciate the feedback. I don't think this needs to be a v1.0 item however I will consider making this change in the future.

> 2. Since this app is meant to run on the command line I'd recommend changing the main file name to something more specific such as duplicate_checker.py

This change is done.

> 3. I notice that 2 files are generated when you finish a hashing run but it looks like they return pretty much the same information. The duplicates file just looks like it's a shorter list of only duplicates. Since your object has a flag of isDuplicate this information could easily be parsed out. I'd recommend returning only a single file and/or giving the user the option to get the full report or just the duplicates. Either way, I think returning a single file may be more clear.

I partially did this, I decided to just make this one file for simplicity.

> 4. After the program runs it only tells the user if there were duplicates or not and doesn't present the users with any further helpful information. I would recommend that if you're going to prompt users for input to add some more functionality after a run has been completed. This could be an option to see in the console the number of duplicates found along with a list of filenames. Additionally, other features such as asking the user if they would like to run another comparison could enhance the application as well.

I added some functionality for this, I added info about the run as output in the console however I'm not doing the reprompt as part of v1.0

ssailor96 committed:
Fixing issue with timezone and json naming

ssailor96 committed:
Merge pull request #5 from ssailor96/fix-to-fit-standards

Fix to fit standards

ssailor96 committed:
README has been fixed

Getting Started section added to README. Clarifications and output details have also been added.

ssailor96 committed:
Merge pull request #6 from ssailor96/fix-readme

README has been fixed

ssailor reviewed (approved):
Looks good, go ahead and merge. Make sure to tag it as the v.1.0.0 release in main.
